### ISSUE TITLE
Hydration failed because the server rendered HTML didn't match the client.

### ISSUE THREAD

**Issue by joelachance** [vercel/speed-insights#89]
When rendering my Next app, I'm getting an SSR error from SpeedInsights. How do I know it's from Speed Insights? I can comment it out and the error goes away.

Here's the error: 
```
react-dom-client.development.js:4100 Uncaught Error: Hydration failed because the server rendered HTML didn't match the client. As a result this tree will be regenerated on the client. This can happen if a SSR-ed Client Component used:

- A server/client branch `if (typeof window !== 'undefined')`.
- Variable input such as `Date.now()` or `Math.random()` which changes each time it's called.
- Date formatting in a user's locale which doesn't match the server.
- External changing data without sending a snapshot of it along with the HTML.
- Invalid HTML tag nesting.

It can also happen if the client has a browser extension installed which messes with the HTML before React loaded.

https://react.dev/link/hydration-mismatch

  ...
    <HotReload assetPrefix="">
      <ReactDevOverlay onReactError={function} state={{nextId:1, ...}} dispatcher={{...}}>
        <DevRootNotFoundBoundary>
          <NotFoundBoundary notFound={<NotAllowedRootNotFoundError>}>
            <NotFoundErrorBoundary pathname="/" notFound={<NotAllowedRootNotFoundError>} notFoundStyles={undefined} ...>
              <RedirectBoundary>
                <RedirectErrorBoundary router={{...}}>
                  <Head>
                  <link>
                  <script>
                  <script>
                  <RootLayout>
                    <html lang="en">
                      <head>
                      <body>
                      <SpeedInsights2>
+                       <Suspense fallback={null}>
                      ...
                  ...
        ...

    at throwOnHydrationMismatch (react-dom-client.development.js:4100:11)
    at updateSuspenseComponent (react-dom-client.development.js:8728:15)
    at beginWork (react-dom-client.development.js:9972:18)
    at runWithFiberInDEV (react-dom-client.development.js:543:16)
    at performUnitOfWork (react-dom-client.development.js:14996:22)
    at workLoopConcurrent (react-dom-client.development.js:14990:9)
    at renderRootConcurrent (react-dom-client.development.js:14965:15)
    at performWorkOnRoot (react-dom-client.development.js:14300:13)
    at performWorkOnRootViaSchedulerTask (react-dom-client.development.js:15852:7)
    at MessagePort.performWorkUntilDeadline (scheduler.development.js:44:48)
```

I'm [seeing some Stack Overflows](https://stackoverflow.com/a/73530036) referencing issues where you might have HTML elements inside one another, like `p` els inside other `p` els, etc. 

[You can silence this error by doing this.](https://react.dev/reference/react-dom/client/hydrateRoot#show-a-dialog-for-recoverable-hydration-mismatch-errors)

**Comment by dimitridumont** [comment]
Did you add the <SpeedInsights /> component in body (after children)?

**Comment by joelachance** [comment]
> Did you add the component in body (after children)?

yes!

**Comment by ishaanrathodd** [comment]
> > Did you add the component in body (after children)?
> 
> yes!

were you able to fix it? ain't getting a fix

**Comment by joelachance** [comment]
@ishaanrathodd no, I didn’t pursue it further. 
@dimitridumont it doesn’t seem this is functional, what can I do to help here?

**Comment by ishaanrathodd** [comment]
> @ishaanrathodd no, I didn’t pursue it further. 
> 
> @dimitridumont it doesn’t seem this is functional, what can I do to help here?

I got my error fixed by making the file a client component by adding "use client" where the error was found. Try to comment out lines which you think might be the culprit and then find a workaround.

**Comment by misterioso013** [comment]
I realized that in my case the blame was on the extensions I used in my browser, they added style in the wrong places, <div> inside <p> ...
Anyway, I'm using the browser's incognito tab to program with next... 
![Captura de tela de 2024-12-28 15-29-54](https://github.com/user-attachments/assets/eb0749b6-9300-4cdd-9155-2acbbe28a6ed)

**Comment by Md-Anamul-Haque** [comment]
# Clarification on suppressHydrationWarning Usage in <body>
I have a use case where I add suppressHydrationWarning in the <body> tag like this:
```

**Comment by Md-Anamul-Haque** [comment]
# Clarification on suppressHydrationWarning Usage in <body> Tag

I have a use case where I add suppressHydrationWarning in the <body> tag like this:
```tsx
<html lang="en" suppressHydrationWarning>
  <body suppressHydrationWarning>
    ...
  </body>
</html>
```

**Comment by Willy-Wonka28** [comment]
> In my case, The Grammarly extension was adding styles in the wrong places and thus the error. Disabling the extension solved the issue.

Hey! I had the same issue and after disabling my Grammarly extension, the error went away. 

Thanks!

**Comment by Willy-Wonka28** [comment]
Ok then. I would love to know what you find out.

**Comment by ViniciusLima7** [comment]
The same error happens to me, even in an anonymous tab

**Comment by Gaurav3697** [comment]
suppressHydrationWarning  is helpful

**Comment by ahmvdev** [comment]
> # Clarification on suppressHydrationWarning Usage in Tag
> I have a use case where I add suppressHydrationWarning in the tag like this:
> 
> <html lang="en" suppressHydrationWarning>
>   <body suppressHydrationWarning>
>     ...
>   </body>
> </html>

this fixed it for me

**Comment by joelachance** [comment]
> If youre working with nextjs, typescript then follow the procedure. Go the layout.tsx ->
> 
> ... add the attached code.

That's what's not working for some folks, myself included @Abir-Zayn

**Comment by fathan** [comment]
its work for me, I implement on this comment https://github.com/vercel/speed-insights/issues/89#issuecomment-2599679417

**Comment by ViniciusLima7** [comment]
suppressing warnings is not a solution, it's just a palliative, we're just hiding the error.

**Comment by aleynaeser** [comment]
Is there a solution for this? When I opened a project from scratch, I got the error. There is only layout and page. suppressHydrationWarning does not seem like a real solution.

**Comment by vladilen11** [comment]
Is suppressHydrationWarning recommended for use under production env?

**Comment by iiiok** [comment]
In my case , I use "process.env.STRAPI_BASE_URL" as a para. 

I guess it is this env para cause the "This won't be patched up. This can happen if a SSR-ed Client Component used" error.

**Comment by frontendDev007** [comment]
I got the same issue I don't use any api and every components/pages is my client side which I use "use client" but also I got error. If you use chrome and some extension on that then this error appears I run same on my brave browser and there is no error.

**Comment by mi-heorhi** [comment]
> > In my case, The Grammarly extension was adding styles in the wrong places and thus the error. Disabling the extension solved the issue.
> 
> Hey! I had the same issue and after disabling my Grammarly extension, the error went away.
> 
> Thanks!

Hmm... will it persist for other users with the Grammarly extension enabled?

**Comment by Alireza9476** [comment]
In my case I had  <button id={`image-${num.index}`} ........
and index was an uuid that was abviously different everytime (lack of knowledge)

const buttonList: numType[] = Array.from({ length: 10 }, (_, i) => ({
  index: uuidv4(),......................
return (..................)

therefore I got this error, fixing it by using another approach
will not fix ur issue but it may help someone else

**Comment by xteqr1** [comment]
In my case, Grammarly and Colozilla extensions were the cause of this problem.

**Comment by fait-arch** [comment]
Este video me ayudo  => [Video](https://www.youtube.com/watch?v=RFf1dkzm-T0)


```
//src\components\ClientRender.tsx
'use client';

import React, { ReactNode, useEffect, useState } from 'react';

export const RenderMounted = ({ children }: { children: ReactNode }) => {
  const [mounted, setMounted] = useState(false);
  useEffect(() => setMounted(true), []);
  if (!mounted) return null;
  return <>{children}</>;
};
```

```
//src\app\layout.tsx
<RenderMounted>
	<ThemeProvider
		attribute="class"
		defaultTheme="light"
		enableSystem={false}
	>
		{children}
	</ThemeProvider>
</RenderMounted>
```



###  Detalles
Aun asi el problema ocurre en el segundo refresco de la página, cuando tengo la extensión ColorZilla instalada. Lo probe en firebox.

**Comment by Abhi-6284** [comment]
"Hi, I have a solution for this.  

Use an empty fragment:  

```jsx
<>
  <html lang="en">
    <body className={`${geistSans.variable} ${geistMono.variable} antialiased`}>
      <Navbar />
      {children}
    </body>
  </html>
</>
```
Hope this helps!"

**Comment by danvoltura** [comment]
> I got the same issue I don't use any api and every components/pages is my client side which I use "use client" but also I got error. If you use chrome and some extension on that then this error appears I run same on my brave browser and there is no error.

>In my case, Grammarly and Colozilla extensions were the cause of this problem.

Colozilla was also the culprit on my end. 

**suppressHydrationWarning** is not really a solution... frontend ocd

**Comment by UselessParadox** [comment]
Same issue this is so crazy .. i sit here since ages. Go through entire tree levels.. its impossible to fix!  That wont fix shit  suppressHydrationWarning  and i don't prefer that. Rather have the issue fixed.

**Comment by joelachance** [comment]
@feugy want to jump in here? This is clearly an issue Vercel should fix.

**Comment by EnettyTech** [comment]
Any solusion? I was used: state mounted, suppressHydrationWarning but not working.

**Comment by joelachance** [comment]
@tobiaslins thanks for the follow up- it seems like it's maybe chrome-extension related. I'm not sure I can get to it until end of next week, but I'll look at it again. This one is interesting at least :)

Thanks again for the response!

**Comment by Smileatmaxi** [comment]
Hey ho,

I joined this conversation because I faced the same issue, and I think I found the root cause. For me, the problem was related to random number generation in a component with "use client".

In my code, I had a function that generated random numbers for the Y-axis rotation:

`const randomRotateY = () => {
    return Math.floor(Math.random() * 21) - 10;
};`

However, because the component is client-side rendered with "use client", this code was executed differently on the server and the client. The random number generation on the server could result in a different value from the client-side, causing inconsistencies and unexpected behavior.

To fix this, I made sure that the random number generation only happens on the client side. Here's how I adjusted the code:

`const [rotation, setRotation] =useState(0);`

`useEffect(() => {
    setRotation(randomRotateY()); 
}, []);` **_// Ensure this only runs on the client_**


By doing this, I ensure that the random number is generated on the client side, and the rotation value is consistently set on the client, not on the server.

This solution fixed the problem for me, and I hope it helps you too 
at least in finding out what your problem comes from!

**Comment by hadnet** [comment]
I use dark mode, so added 'dark' to the html tag (<html className="dark">) and no more hydration warning.

**Comment by AIABHISHEK** [comment]
> Este video me ayudo => [Video](https://www.youtube.com/watch?v=RFf1dkzm-T0)
> 
> ```
> //src\components\ClientRender.tsx
> 'use client';
> 
> import React, { ReactNode, useEffect, useState } from 'react';
> 
> export const RenderMounted = ({ children }: { children: ReactNode }) => {
>   const [mounted, setMounted] = useState(false);
>   useEffect(() => setMounted(true), []);
>   if (!mounted) return null;
>   return <>{children}</>;
> };
> ```
> 
> ```
> //src\app\layout.tsx
> <RenderMounted>
> 	<ThemeProvider
> 		attribute="class"
> 		defaultTheme="light"
> 		enableSystem={false}
> 	>
> 		{children}
> 	</ThemeProvider>
> </RenderMounted>
> ```
> 
> ### Detalles
> Aun asi el problema ocurre en el segundo refresco de la página, cuando tengo la extensión ColorZilla instalada. Lo probe en firebox.

Worked for me too

**Comment by tainv1101** [comment]
My solution is to use dynamic from next/dynamic for the component in the library that's causing this error.
`"use client";
import dynamic from "next/dynamic";
import { ResponsiveMasonry } from "react-responsive-masonry";
const Masonry = dynamic(() => import("react-responsive-masonry"), { ssr: false });`

**Comment by tainv1101** [comment]
My solution is to use dynamic from next/dynamic for the component in the library that's causing this error.
`"use client";`
`import dynamic from "next/dynamic";`
`import { ResponsiveMasonry } from "react-responsive-masonry";`
`const Masonry = dynamic(() => import("react-responsive-masonry"), { ssr: false });`

**Comment by joelachance** [comment]
> My solution is to use dynamic from next/dynamic for the component in the library that's causing this error. `"use client";` `import dynamic from "next/dynamic";` `import { ResponsiveMasonry } from "react-responsive-masonry";` `const Masonry = dynamic(() => import("react-responsive-masonry"), { ssr: false });`

What's masonry? Any idea why this works for you @tainv1101?

**Comment by GulsahDuzgun** [comment]
l have been struggling with same issue. But l found it causes from that we try to map our variable with same time Next.js was hydrating the page.  To prevent that l  

 const [isClientSide, setIsClientSide] = useState(false);

  useEffect(() => {
    setIsClientSide(true);
  }, []);

And then in other useEffect hook -- that l define default values of my form inputs that l use with useForm() -- l give that value in array like 
 useEffect(() => {
    if (currentFacilityDetail && isClientSide) {
      reset({
      name:currentFacilityDetail.name
      });
    }
  }, [currentFacilityDetail, reset, isClientSide]);
 
to assign variables after created page layout and theme classes

**Comment by Wuzoobia** [comment]
If you have **Grammarly Chrome extension**, just disable it. If that didn't work,
The other solution is to add suppressHydrationWarning to **both html and body tags**, this will prevent hydration warnings from bubbling up from child components. Also need to ensure any dynamic content is properly handled with 'useEffect'

**Comment by monteiz** [comment]
There are plenty of reasons this error can come up, so it is useless to find a universal solution (and suppressing the warning is a very poor software decision, always).

This error shows up because there is a difference between the server side generated content, and the client side.

The right thing to do is checking the browser console, that will highlight where exactly this difference is.

From there, you can likely see which component of your app is creating the conflict, and fix it precisely.

**Comment by AIABHISHEK** [comment]
> There are plenty of reasons this error can come up, so it is useless to seek for a universal solution (and suppressing the warning is a very poor design decision, always).
> 
> This error shows up because there is a difference between the server side generated content, and the client side.
> 
> The right thing to do is checking the browser console, that will highlight where exactly this difference is.
> 
> From there, you can likely see which component of your app is creating the conflict, and fix it precisely.

Yes later I figured it out and able to solve it without any work-around.

**Comment by joelachance** [comment]
> There are plenty of reasons this error can come up, so it is useless to seek for a universal solution (and suppressing the warning is a very poor design decision, always).

It's possible there's a universal solution here, it's worth digging into to see if one exists.

**Comment by monteiz** [comment]
@joelachance the source of this error can be any of your app component, function, or browser extension.

Apart from manually suppressing the warning itself (which is not a solution), hoping to find a universal solution is not reasonable.

This issue should be closed.

**Comment by joelachance** [comment]
> the source of this error can be any of your app component, function, or browser extension.

Until you can prove that, @monteiz , this issue should stay open. Opinions are appreciated, but the point here is to get to root cause, not assumptions.

**Comment by monteiz** [comment]
@joelachance Yeah, you're right, this issue should not be closed. I forgot that you made it clear that this issue is specific to SpeedInsights, which in your case is the precise source of the problem. My bad.

**Comment by Cisco0911** [comment]
Hi everyone,

I'm encountering a similar issue in my Next.js App Router setup and wanted to provide additional context in case it's helpful.

In my case, the error appears when all of the following conditions are true:

page.tsx is an async server component (e.g., await new Promise(...))

There's a loading.tsx file in the same route

A 'use client' component is imported and rendered within page.tsx (even without using hooks)

The browser DevTools are open during development

I make a hot reload change (e.g., save the client component or loading file)


After hot reloading, refreshing the page results in hydration errors such as:

`Uncaught Error: Hydration failed because the server rendered text didn't match the client...`

Or sometimes:

`Invalid hook call. Hooks can only be called inside of the body of a function component.`

If a hook is used in the client component.

I'm just not sure whether I'm doing something wrong or should I just ignore it with `suppressHydrationWarning` ?

Thank you for your help.

**Comment by rohDevellopper** [comment]
If you have a website or app and you want to index it, I recommend not silencing the SSR because doing so would affect SEO.

**Comment by YasinKuralay** [comment]
> EDIT: I noticed the error only occurs with MS Edge and Chrome DevTools, not with Firefox. Maybe it is related to chronium...

Can confirm, I have the same problem. Only happens in chromium (chrome, electron-cypress) and not in firefox.

**Comment by kelivdc** [comment]
I found solution for this problem by following this 
https://mui.com/material-ui/integrations/nextjs/

I just only change the layout.js like this
<html lang="en" >
      <AppRouterCacheProvider>
      <body
        className={`antialiased ${open_sans.className}`}
        >
        {children}
      </body>
        </AppRouterCacheProvider>
    </html>

This is solved hydration error

**Comment by Pranav0728** [comment]
> > [@ishaanrathodd](https://github.com/ishaanrathodd) no, I didn’t pursue it further.
> > [@dimitridumont](https://github.com/dimitridumont) it doesn’t seem this is functional, what can I do to help here?
> 
> I got my error fixed by making the file a client component by adding "use client" where the error was found. Try to comment out lines which you think might be the culprit and then find a workaround.

Thanks it helps me 👍

**Comment by DavidSouzaLD** [comment]
I decided to do it this way, even though I don't know if it's the best option.

```
"use client"
import { useState, useEffect } from "react"

function Example()
{
const [isClientSide, setIsClientSide] = useState(false);

    useEffect(() => {
        if (!isClientSide)
            setIsClientSide(true);
    }, []);


return(
     {isClientSide && items.map((item) => ()} // Problem section
)
}
```

Let me know if this helped you too.

**Comment by Roter-S** [comment]
In my case, disabling the "Urban VPN Proxy" browser extension solved the problem.

**Comment by kareem-cureocity** [comment]
In my case, a wrapped children component with "Suspense" , got issue fixed

` <Suspense
          fallback={
            <div className="flex min-h-screen flex-col items-center justify-center gap-3 bg-background">
              <Spinner />
              <p className="text-foreground">Loading...</p>
            </div>
          }
        >
          <Providers>
            {children}
            <Toaster
              position="top-right"
              richColors
              closeButton
              duration={4000}
            />
          </Providers>
        </Suspense>`

**Comment by petrsonTD** [comment]
Guys I had big brain moment. I also had this problem with NextJS only God knows how long and with every NextJS project. I recently started new NextJS project and again had this problem. Event tho there was almost nothing in it. So I removed every component and I had just basic html elements in index. I didn't  have error bud my styles were off. Only when I made hard refresh ( command + shift + r ), they loaded, this also temporarily fixed problem with hydration, well till I made normal refresh. So obviously it was problem with cache. As few of you mentioned, if you use different browser (safari for example), problem doesn't occur.

Long story short, SOLUTION for me was in chrome go to Settings, and in left menu select "Reset settings" and click "Restore settings to their original defaults". Hope this helps.

**Comment by Anurag-NURA** [comment]
Might be stupid to say it but i added head tag before body tag and it removed the hydration mismatch problem. I am using next 15.3.4
